### ISSUE TITLE
Possible bug with initial sizing of image layout

### ISSUE THREAD

**Issue by evanmanning** [schlosser/pig.js#15]
First of all, I wanted to say that you've done an amazing job with this library; I think it's fantastic. Of the libraries that I've looked into for displaying images in a tiled fashion, this is the one that has met my needs, especially since it displays them in rows (like in Google Photos), which I really like.

I've noticed what might be a bug, so I'll do my best to describe it. Suppose you have a page with a `<div id="pig">` to hold the pig layout, as usual when using the library. I've noticed that when the page is first loaded, the image layout as a whole seems to be scaled a little too big, such that it oversteps the right-hand boundary of the div. One way to notice this is to look at a page that has left and right margins (or padding) of equal width on the sides of the content: when the page is first loaded, the right margin will appear a little narrower because the image layout is overstepping the boundary somewhat on the right side. However, if you resize the browser window after the initial page load, everything seems to get sized correctly (and, in this example, the margins would now appear equal).

I was able to observe this phenomenon in Chrome, Firefox, and Safari (all on desktop), both with a site I'm making, as well as with the [https://feeding.schlosser.io/](https://feeding.schlosser.io/) site (which, by the way, is really cool—I love how you documented all that food!).

Are you able to observe this phenomenon as well? I'd love to help out with a fix if I'm able, but since I relatively inexperienced with JS (I've mostly done back-end work in Java thus far), and since I'm not yet very familiar with this library's implementation, I would probably need some pointers along the way.

In any case, thanks again for your great work!

**Comment by evanmanning** [comment]
Probably the most notable thing I've found so far is the following. I placed statements to log the width of the pig container (like `console.log($('#pig')[0].clientWidth);`) before and after `new Pig(imageData, options).enable();`, and I tried loading the page. I noticed that the container width value before initializing the Pig was larger than the value after, by a difference of 15 (it seems to always be a difference of 15, no matter the browser window width). The smaller value is the "real" one; that is, if you inspect the container element after the page is loaded, its width is the smaller value—and thus the smaller value is also the one that is correct in terms of appearance. This helps explain a proximate cause of the overstepping of the boundary, because it seems that the initial `_computeLayout()` is done using a container width that's 15 larger than it should be, for whatever reason. Of course, the ultimate cause for the width discrepancy is still a mystery.

I drilled down into `enable()` to check the values in a finer-grained fashion, like so:

```js
  Pig.prototype.enable = function() {
    console.log('1: ' + $('#pig')[0].clientWidth);
    this.onScroll = this._getOnScroll();
    window.addEventListener('scroll', this.onScroll);
    console.log('2: ' + $('#pig')[0].clientWidth);
    this.onScroll();
    console.log('3: ' + $('#pig')[0].clientWidth);
    this._computeLayout();
    console.log('4: ' + $('#pig')[0].clientWidth);
    this._doLayout();
    console.log('5: ' + $('#pig')[0].clientWidth);
    optimizedResize.add(function() {
      this.lastWindowWidth = window.innerWidth;
      this._computeLayout();
      this._doLayout();
    }.bind(this));
    console.log('6: ' + $('#pig')[0].clientWidth);
    return this;
  };
```
It appears that the change in width happens during `this._doLayout();`, which seems reasonable, given that the DOM manipulations happen there. This is what I've found so far, and I hope to drill into the `_doLayout()` function soon.

To make things more interesting, it seems that this issue does not occur when using mobile Chrome on Android or the mobile simulator in the developer tools in desktop Chrome (those are the only mobile browsers I've tested so far).

**Comment by schlosser** [comment]
Thanks for reporting this issue! Unfortunately, I can't reproduce it on my computer. A few questions:

1. What OS are you seeing this on? 
2. Can you take a screenshot of this? Or is it only for a second? Can you take a video?

Thanks,
Dan

P.S. Glad you're enjoying the library!!

**Comment by evanmanning** [comment]
Thanks for your response, and I apologize for taking so long to reply.

I'm using OS X 10.11.6 (El Capitan) and Google Chrome 63.0.3239.132, although I see the issue with Firefox and Safari as well.

Hopefully these images help illustrate the phenomenon. As you can probably tell, the gray bars on either side are just my desktop background, behind the browser window.

State after initial page load:
<img width="728" alt="screen shot 2018-01-17 at 7 08 27 pm" src="https://user-images.githubusercontent.com/15351078/35077866-7c0f2238-fbbc-11e7-8bd1-9ac0de69c463.png">

State after resizing the browser window by any amount:
<img width="724" alt="screen shot 2018-01-17 at 7 08 51 pm" src="https://user-images.githubusercontent.com/15351078/35077901-b3eb5f5a-fbbc-11e7-88cd-19274d5173e9.png">

You'll notice that initially, the margin on the right side of the layout (pointed to by the red arrow) is smaller than the corresponding margin on the left side. After resizing the window by any amount, the right margin becomes the correct size, equal in width to the left margin.

I was able to find an interesting workaround hack. In the `Pig.prototype.enable` function, [lines 663–664 in this version](https://github.com/schlosser/pig.js/blob/cd4af51987caba432e2765866556860185217fa4/src/pig.js#L663) do the following:

```java
this._computeLayout();
this._doLayout();
```

By simply repeating these two calls, such that lines 663–666 now look like

```java
this._computeLayout();
this._doLayout();
this._computeLayout();
this._doLayout();
```

the issue goes away. Definitely not elegant, and I don't fully understand why it works, but at least it's solving the issue for me at the moment. : )

I don't know if this should be considered a bug or just some unfortunate consequence of the combination of myriad variables on my machine, but hopefully the above might be interesting for what it's worth.

**Comment by schlosser** [comment]
A-ha! I'm able to reproduce this now, thanks for the detailed instructions.

I agree that your solution is inelegant, but it's helpful to know that it works. This seems like a race condition of some sort, I'll take a look.

**Comment by schlosser** [comment]
Fixed: this bug is due to the fact that the pages you were testing on had less than 1 full page worth of content, so the vertical scrollbar was not present. When the images rendered for the first time, the scrollbar was added (after the fact, and the layout wasn't recomputed).

I added CSS styling on the body to ensure the scrollbar is shown even before the first render.

https://github.com/schlosser/pig.js/commit/7814a7e3f4235c9cf149bb84b688b58eb48a872d
